### PR TITLE
dispatch new events after line item has been added / removed / quanti…

### PR DESCRIPTION
**Pull request by maqavelli**
### 1. Why is this change necessary?
The currently available LineItemAddedEvent is the state before the line item is saved to the cart, which in combination with the rule builder, does not return the ID of the rule in the sales channel context in a subscriber to that event.
   
In order to obtain the rule ID, the event must be dispatched after the cart has been saved.
   
Therefor it's suggested to create new events `LineItemAddedFinishEvent`, `LineItemRemovedFinishEvent` and `LineItemQuantityChangedFinishEvent`.

### 2. What does this change do, exactly?
Dispatch new events after the line item has been saved to the cart

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/master/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: src/Core/Checkout/Cart/Event/LineItemAddedFinishEvent.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Checkout\Cart\Event;
+
+use Shopware\Core\Checkout\Cart\Cart;
+use Shopware\Core\System\SalesChannel\SalesChannelContext;
+use Symfony\Contracts\EventDispatcher\Event;
+
+/**
+ * @deprecated tag:v6.4.0 - Will implement Shopware\Core\Framework\Event\ShopwareSalesChannelEvent
```
Comment (mitelg) [inline src/Core/Checkout/Cart/Event/LineItemAddedFinishEvent.php]: copy paste :smile: 

there is no need to do this on new events. please remove the deprecations completely and implement the interface correctly

File: src/Core/Checkout/Cart/Event/LineItemAddedFinishEvent.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    public function getCart(): Cart
+    {
+        return $this->cart;
+    }
+
+    /**
+     * @deprecated tag:v6.4.0 - Will return Shopware\Core\Framework\Context instead
+     */
+    public function getContext(): SalesChannelContext
```
Comment (mitelg) [inline src/Core/Checkout/Cart/Event/LineItemAddedFinishEvent.php]: this return type is against the definition from the interface. this should break on your machine right now :blush: 

please also remove the deprecation message

**Review by philipreinken — CHANGES_REQUESTED** [review]
Hello @maqavelli , thank you for creating this pull request :slightly_smiling_face: I only have a few suggestions here. Since we use this terminology in the core already, try to rename the events using before/after terminology, for Example: `BeforeSomethingHappened` and `AfterSomethingHappened`. Also, changelog entries are currently written using a certain [pattern](https://github.com/shopware/platform/blob/master/adr/2020-08-03-Implement-New-Changelog.md#workflow), please adhere to this (the reasoning behind your changes could for example be moved to the pull request description). Apart from that, please add tests for your changes and I think we're good to go :rocket:

**Comment by maqavelli** [comment]
@philipgatzka I've made the changes as you requested. Some of the tests failed, but these were not created by myself.

**Comment by philipreinken** [comment]
Nice :ok_hand: I don't think all the test failures are unrelated though. These two PHPUnit tests could either have wrong assertions, or the events are fired too often indeed:

```
There were 2 failures:

1) Shopware\Core\Checkout\Test\Cart\CartServiceTest::testAfterLineItemQuantityChangedEventFired
Shopware\Core\Framework\Test\TestCaseHelper\CallableClass::__invoke(Shopware\Core\Checkout\Cart\Event\LineItemQuantityChangedEvent Object (...), 'Shopware\Core\Checkout\Cart\E...dEvent', Symfony\Component\HttpKernel\Debug\TraceableEventDispatcher Object (...)): void was not expected to be called more than once.

/__w/platform/platform/development/vendor/symfony/event-dispatcher/Debug/WrappedListener.php:126
/__w/platform/platform/development/vendor/symfony/event-dispatcher/EventDispatcher.php:251
/__w/platform/platform/development/vendor/symfony/event-dispatcher/EventDispatcher.php:73
/__w/platform/platform/development/platform/src/Core/Framework/Webhook/WebhookDispatcher.php:88
/__w/platform/platform/development/platform/src/Core/Framework/Event/BusinessEventDispatcher.php:46
/__w/platform/platform/development/platform/src/Core/Framework/Event/NestedEventDispatcher.php:32
/__w/platform/platform/development/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php:168
/__w/platform/platform/development/platform/src/Core/Checkout/Cart/LineItemFactoryRegistry.php:86
/__w/platform/platform/development/platform/src/Core/Checkout/Cart/SalesChannel/CartItemUpdateRoute.php:78
/__w/platform/platform/development/platform/src/Core/Checkout/Cart/SalesChannel/CartService.php:164
/__w/platform/platform/development/platform/src/Core/Checkout/Test/Cart/CartServiceTest.php:263

2) Shopware\Core\Checkout\Test\Cart\LineItemFactoryRegistryTest::testChangeQuantity
Shopware\Core\Framework\Test\TestCaseHelper\CallableClass::__invoke(Shopware\Core\Checkout\Cart\Event\LineItemQuantityChangedEvent Object (...), 'Shopware\Core\Checkout\Cart\E...dEvent', Symfony\Component\HttpKernel\Debug\TraceableEventDispatcher Object (...)): void was not expected to be called more than once.

/__w/platform/platform/development/vendor/symfony/event-dispatcher/Debug/WrappedListener.php:126
/__w/platform/platform/development/vendor/symfony/event-dispatcher/EventDispatcher.php:251
/__w/platform/platform/development/vendor/symfony/event-dispatcher/EventDispatcher.php:73
/__w/platform/platform/development/platform/src/Core/Framework/Webhook/WebhookDispatcher.php:88
/__w/platform/platform/development/platform/src/Core/Framework/Event/BusinessEventDispatcher.php:46
/__w/platform/platform/development/platform/src/Core/Framework/Event/NestedEventDispatcher.php:32
/__w/platform/platform/development/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php:168
/__w/platform/platform/development/platform/src/Core/Checkout/Cart/LineItemFactoryRegistry.php:86
/__w/platform/platform/development/platform/src/Core/Checkout/Test/Cart/LineItemFactoryRegistryTest.php:118

ERRORS!
Tests: 1620, Assertions: 5027, Errors: 9, Failures: 2, Skipped: 23.
```

Also, the failing Cypress test may have something to do with this as well, since it fails when adding line items:

```
1) Order: Create order
       @base @order: create order with a new customer, update line item and shipping cost manually:

      AssertionError: expected { Object (xhr, id, ...) } to have property 'status' of 200, but got 500
      + expected - actual

      -500
      +200
```

Please have a look at the failing tests and make sure they're indeed unrelated. I'd recommend running the tests locally (at least PHPUnit) for this, since it will be quicker to debug this way.

**Comment by maqavelli** [comment]
@philipgatzka Hi Philip all tests have successfully passed

**Review by philipreinken — APPROVED** [review]
Nice, thanks for having another go :+1:

File: src/Storefront/Controller/CartLineItemController.php [hunk 3]
```diff
@@ -3,6 +3,11 @@
 namespace Shopware\Storefront\Controller;
 
 use Shopware\Core\Checkout\Cart\Cart;
+use Shopware\Core\Checkout\Cart\Event\CartLineItemAddedEvent;
```
Comment (philipreinken) [inline src/Storefront/Controller/CartLineItemController.php]: Firing the events in the controller is superfluous, since this has been built with the [store API](https://docs.shopware.com/en/shopware-platform-dev-en/developer-guide/controller#store-api-controller-route). This means every service call here in turn calls the store API route and therefore already triggers the events.

File: src/Core/Checkout/Cart/Event/AfterLineItemAddedEvent.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
+<?php declare(strict_types=1);
```
Comment (philipreinken) [inline src/Core/Checkout/Cart/Event/AfterLineItemAddedEvent.php:1]: Thank you for updating the naming scheme - since we need this to be consistent though, please change the events to e.g. "BeforeLineItemAddedEvent" and "AfterLineItemAddedEvent". The existing "LineItemAddedEvent" can then be flagged for removal (`@deprecated ...`) with the upcoming major version v6.4.0 (it already is marked as deprecated, due to an interface change).

**Review by philipreinken — CHANGES_REQUESTED** [review]
Hi @maqavelli ,

this has been reviewed internally again, there are a few minor problems I've missed the last time, would you mind updating your code once more?

File: src/Core/Checkout/Cart/Event/BeforeLineItemAddedEvent.php [hunk 5]
```diff
@@ -10,7 +10,7 @@
 /**
  * @deprecated tag:v6.4.0 - Will implement Shopware\Core\Framework\Event\ShopwareSalesChannelEvent
  */
-class LineItemAddedEvent extends Event /*implements ShopwareSalesChannelEvent*/
+class BeforeLineItemAddedEvent extends Event /*implements ShopwareSalesChannelEvent*/
```
Comment (mitelg) [inline src/Core/Checkout/Cart/Event/BeforeLineItemAddedEvent.php]: this is a break. please create a new event class and throw both events first. also add a deprecation message, for the event, which should not be dispatched anymore

File: src/Core/Checkout/Cart/LineItemFactoryRegistry.php [hunk 6]
```diff
@@ -83,7 +83,7 @@ public function update(Cart $cart, array $data, SalesChannelContext $context): v
         if (isset($data['quantity'])) {
             $lineItem->setQuantity($data['quantity']);
 
-            $this->eventDispatcher->dispatch(new LineItemQuantityChangedEvent($lineItem, $cart, $context));
+            $this->eventDispatcher->dispatch(new BeforeLineItemQuantityChangedEvent($lineItem, $cart, $context));
```
Comment (mitelg) [inline src/Core/Checkout/Cart/LineItemFactoryRegistry.php:89]: the old events still need to be dispatched. otherwise this would be breaking for plugins subscribing to those events

**Review by philipreinken — CHANGES_REQUESTED** [review]
Hi @maqavelli , thank you for the update. There are however still events which are [dispatched in the controller](https://github.com/shopware/platform/pull/1494#discussion_r549229276), please remove those calls. Apart from that, please update your changelog [according to our contribution guideline](https://github.com/shopware/platform/blob/master/adr/2020-08-03-Implement-New-Changelog.md#changelog-entries).

File: src/Storefront/Controller/CartLineItemController.php [hunk 7]
```diff
@@ -51,16 +52,23 @@ class CartLineItemController extends StorefrontController
      */
     private $productLineItemFactory;
 
+    /**
+     * @var EventDispatcherInterface
+     */
+    private $eventDispatcher;
```
Comment (mitelg) [inline src/Storefront/Controller/CartLineItemController.php]: this is not used anymore

File: src/Core/Checkout/Cart/SalesChannel/AbstractCartItemAddRoute.php [hunk 8]
```diff
@@ -13,5 +13,5 @@ abstract class AbstractCartItemAddRoute
 {
     abstract public function getDecorated(): AbstractCartItemAddRoute;
 
-    abstract public function add(Request $request, Cart $cart, SalesChannelContext $context, ?array $items): CartResponse;
+    abstract public function add(Request $request, Cart $cart, SalesChannelContext $context, ?array $items, bool $dispatchAfter): CartResponse;
```
Comment (mitelg) [inline src/Core/Checkout/Cart/SalesChannel/AbstractCartItemAddRoute.php]: why did you add this parameter everywhere? what is the purpose of this?

Comment (maqavelli) [reply src/Core/Checkout/Cart/SalesChannel/AbstractCartItemAddRoute.php]: If you subscribe to the event `AfterLineItemAddedEvent` and the subscriber is adding a gift product after a specific product is added to the cart it needs a flag to prevent the event being fired again or it ends in an endless loop.

Comment (mitelg) [reply src/Core/Checkout/Cart/SalesChannel/AbstractCartItemAddRoute.php]: to prevent this, you should add an extension to the `Context` object and check for this extension in your subscriber

besides that, adding a new required parameter to a method is a breaking change :wink:

Comment (maqavelli) [reply src/Core/Checkout/Cart/SalesChannel/AbstractCartItemAddRoute.php]: The parameter has a default value to make it optional. My subscriber to add the gift is using the CartService to add it to the cart and by using this class it goes over the CartItemAddRoute, so there is no way to prevent it in a subscriber, it has to be done in the CartItemAddRoute.

Comment (mitelg) [reply src/Core/Checkout/Cart/SalesChannel/AbstractCartItemAddRoute.php]: you said, you are in an endless loop because you subscribe to the same event which would be fired in your subscriber. But:
1. this would affect all events and not only the `After...` events :wink: 
2. you could simply do the following:
    - event subscriber method is called 
    - get the `\Shopware\Core\Framework\Context` from the given *Event object 
    - add an extension to the context object (`\Shopware\Core\Framework\Struct\ExtendableTrait::addExtension`)
    - add your new item to the cart
    - event subscriber method is called again, because you add an item to the cart
    - check if the Context object contains your extension. if so, remove the extension from the context, return and do nothing. if it does not contain your extension, your code will be executed normally

Comment (maqavelli) [reply src/Core/Checkout/Cart/SalesChannel/AbstractCartItemAddRoute.php]: Ah now I see it :D let me try it quickly, then I'll remove the parameter

Comment (maqavelli) [reply src/Core/Checkout/Cart/SalesChannel/AbstractCartItemAddRoute.php]: I've made the changes, but I noticed If anyone has multiple plugins subscribing to the AfterLineItemAddedEvent and there is one plugin doing this add gift, all of these subscribers needs to check if the context has the extension and return the function. So by using the parameter to not dispatch the events seems better to me, but If you insist to have it like that, I leave it to this current state.

Comment (philipreinken) [reply src/Core/Checkout/Cart/SalesChannel/AbstractCartItemAddRoute.php]: Yes that's fine, any subscriber will need to take these precautions if it works similarly to yours. Those other subscribers would have their own context extension to check for.

File: src/Storefront/DependencyInjection/controller.xml [hunk 9]
```diff
@@ -75,6 +75,7 @@
             <argument type="service" id="Shopware\Core\Content\Product\Cart\ProductLineItemFactory"/>
         </service>
 
+
```
Comment (mitelg) [inline src/Storefront/DependencyInjection/controller.xml]: could you revert this unnecessary change?

File: src/Core/Checkout/Cart/LineItemFactoryRegistry.php [hunk 10]
```diff
@@ -84,6 +85,7 @@ public function update(Cart $cart, array $data, SalesChannelContext $context): v
             $lineItem->setQuantity($data['quantity']);
 
             $this->eventDispatcher->dispatch(new LineItemQuantityChangedEvent($lineItem, $cart, $context));
```
Comment (mitelg) [inline src/Core/Checkout/Cart/LineItemFactoryRegistry.php:88]: @philipgatzka do you think, we need a deprecation message here, too? that this event dispatching will be removed? or is it sufficient, that the event class is deprecated?

**Review by mitelg — CHANGES_REQUESTED** [review]

File: src/Core/Checkout/Cart/LineItemFactoryRegistry.php (hunk 10 above)
Comment (philipreinken) [reply src/Core/Checkout/Cart/LineItemFactoryRegistry.php:88]: Good point, another deprecation comment would be nice to have I think, since it would be directly visible to developers independent of the IDE (-features) they're using.
```suggestion
            /** @deprecated tag:v6.4.0 - The LineItemQuantityChangedEvent will be removed in the future, please use the BeforeLineItemQuantityChangedEvent and AfterLineItemQuantityChangedEvent variants of this event going forward */
            $this->eventDispatcher->dispatch(new LineItemQuantityChangedEvent($lineItem, $cart, $context));
```

**Review by mitelg — APPROVED** [review]

**Review by philipreinken — APPROVED** [review]

**Comment by philipreinken** [comment]
Thank you for contributing @maqavelli ! 🥳
